### PR TITLE
Documentation-AM-3.20

### PR DESCRIPTION
**Pull request by leleueri**
https://gravitee.atlassian.net/jira/software/c/projects/AM/boards/5?modal=detail&selectedIssue=AM-205

[AM-205]: https://gravitee.atlassian.net/browse/AM-205?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://graviteedocs.blob.core.windows.net/am-3-20-0/index.html)

### CODE REVIEW

File: _data/sidebars/am_3_x_sidebar.yml [hunk 1]
```diff
@@ -426,7 +426,10 @@ entries:
                     url: /am/current/am_userguide_mfa_recovery_codes.html
                   - title: Reset MFA
                     output: web
-                    url: /am/current/am_userguide_mfa_user_factors.html
+                    url: /am/current/am_userguide_mfa_user_factors.
```
Comment (AndrewMcFarlandCampbellGravitee) [inline _data/sidebars/am_3_x_sidebar.yml:429]:
```suggestion
                    url: /am/current/am_userguide_mfa_user_factors.html
```

File: pages/am/3.x/breaking-changes/breaking-changes.adoc [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
 This section outlines the changes you should consider when migrating your application from one version of Gravitee.io Access Management to another.
 
+* link:{{ '/am/current/am_breaking_changes_3.20.html' | relative_url }}[Breaking changes in 3.20]
```
Comment (AndrewMcFarlandCampbellGravitee) [inline pages/am/3.x/breaking-changes/breaking-changes.adoc:9]: This link should also be added to `_data/sidebars/am_3_x_sidebar.yml`

File: am/current/management-api/3.20/swagger.json [hunk 3]
```diff
@@ -0,0 +1 @@
...string"}}},"AuditEntity":{"type":"object","properties":{"id":{"type":"string"},"alternativeId":{"type":"string"},"type":{"type":"string"},"displayName":{"type":"string"},"referenceType":{"type":"string","enum":["PLATFORM","DOMAIN","APPLICATION","ORGANIZATION","ENVIRONMENT"]},"referenceId":{"type":"string"},"attributes":{"type":"object","additionalProperties":{"type":"object"}}}},"AuditOutcome":{"type":"object","properties":{"status":{"type":"string"},"message":{"type":"string"}}},"ScopeApprovalEntity":{"type":"object","properties":{"id":{"type":"string"},"transactionId":{"type":"string"},"userId":{"type":"string"},"clientId":{"type":"string"},"domain":{"type":"string"},"scope":{"type":"string"},"status":{"type":"string","enum":["APPROVED","DENIED"]},"expiresAt":{"type":"integer","format":"int64"},"createdAt":{"type":"integer","format":"int64"},"updatedAt":{"type":"integer","format":"int64"},"clientEntity":{"$ref":"#/definitions/ApplicationEntity"},"scopeEntity":{"$ref":"#/definitions/ScopeEntity"}}},"ScopeEntity":{"type":"object","properties":{"id":{"type":"string"},"key":{"type":"string"},"name":{"type":"string"},"description":{"type":"string"}}},"Credential":{"type":"object","pro
```
Comment (AndrewMcFarlandCampbellGravitee) [inline am/current/management-api/3.20/swagger.json:1]: Should there be some whitespace in this file?

File: _data/sidebars/am_3_x_sidebar.yml (hunk 1 above)
Comment (ashraf706) [reply _data/sidebars/am_3_x_sidebar.yml]: fixed the url

File: pages/am/3.x/breaking-changes/breaking-changes.adoc (hunk 2 above)
Comment (ashraf706) [reply pages/am/3.x/breaking-changes/breaking-changes.adoc:9]: added the missing link

File: am/current/management-api/3.20/swagger.json (hunk 3 above)
Comment (ashraf706) [reply am/current/management-api/3.20/swagger.json:1]: this is a auto generated file and historically we haven't added whitespaces.  We can ignore it.

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+Here is a screenshot of the challenge step with exhausted rate limit:
+
+image::{% link images/am/current/graviteeio-am-userguide-mfa-ratelimit-exceed.png %}[250, 500]
+
+NOTE: You can customise the error message by setting `mfa_challenge.rate.limit.error` property value in the *messages_en.properties* or *messages_fr.properties* file.
+
+
+== Brute Force detection
+
+Brute force is a feature which helps to configure and limit the number of verification requests a user is allowed to send in a specific time period.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:60]:
```suggestion
The Brute Force Detection feature enables you to configure and limit the number of verification requests a user is allowed to send within a specific time period.
```

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+This feature is configurable in domain or in application level. Follow the steps below to configure the *Brute Force* feature in domain level
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:61]:
```suggestion
You can configure Brute Force Detection at domain level or at application level. Follow the steps below to configure the *Brute Force* feature at domain level.
```

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+. link:{{ '/am/current/am_userguide_authentication.html' | relative_url }}[Log in to AM Console^].
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:63]:
```suggestion
. link:{{ '/am/current/am_userguide_authentication.html' | relative_url }}[Log in to the AM Console^].
```

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+. Select *User Accounts* under *SECURITY*.
+. Enable the *Brute Force Detection* in the *MFA* section.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:65]:
```suggestion
. Enable *Brute Force Detection* in the *MFA* section.
```

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+. Define your *Brute Force Detection* preferences.
++
+image::{% link images/am/current/graviteeio-am-userguide-mfa-security.png %}[]
+
+The user will be notified with an error message once the maximum attempt is exhausted.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:70]:
```suggestion
Once the number of maximum attempts is reached, the user will be notified with an error message.
```

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+image::{% link images/am/current/graviteeio-am-userguide-mfa-brute-max.png %}[250, 500]
+
+A new log event `MFA_VERIFY_LIMIT_EXCEED` is also introduced to log the brute force attempt.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:74]:
```suggestion
A new log event, `MFA_VERIFY_LIMIT_EXCEED`, is available to log brute force attempts.
```

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+NOTE: The settings can be overridden at the application level *App > Settings > accounts*.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:76]:
```suggestion
NOTE: You can override domain-level Brute Force Detection settings at the application level from *App > Settings > accounts*.
```

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+You can customise the error message by setting `mfa_challenge.verify.limit.error` property value in the *messages_en.properties* or *messages_fr.properties* file.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:77]:
```suggestion
To customize the error message, modify the value of the `mfa_challenge.verify.limit.error` property in the *messages_en.properties* or *messages_fr.properties* file.
```

File: pages/am/3.x/user-guide/user-management/password/user-guide-password-policy.adoc [hunk 12]
```diff
@@ -23,12 +23,14 @@ NOTE: You can also configure password policy settings at security domain level,
 
 You can set the following password characteristics:
 
-- *Minimum length*: length of the passwords should be enforced by the application.
+- *Minimum & Maximum length*: length of the passwords should be enforced by the application.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/user-management/password/user-guide-password-policy.adoc:26]:
```suggestion
- *Minimum & Maximum length*: length of the passwords that should be enforced by the application.
```

File: pages/am/3.x/user-guide/user-management/password/user-guide-password-policy.adoc [hunk 13]
```diff
@@ -23,12 +23,14 @@ NOTE: You can also configure password policy settings at security domain level,
...
 - *Numbers*: must include at least one number.
+- *Expiry Duration*: the expiration duration (in days) of a password.
 - *Special characters*: must include at least one special character.
 - *Mixed case*: must include lower and upper case letters.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/user-management/password/user-guide-password-policy.adoc:30]:
```suggestion
- *Mixed case*: must include lowercase and uppercase letters.
```

File: pages/am/3.x/user-guide/user-management/password/user-guide-password-policy.adoc [hunk 14]
```diff
@@ -23,12 +23,14 @@ NOTE: You can also configure password policy settings at security domain level,
...
 - *Exclude common passwords*: will exclude common passwords from a dictionary.
 - *Exclude user profile information from passwords*: will user profile information within passwords (case insensitive).
+- *History*: prevent the usage of old password.
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/user-management/password/user-guide-password-policy.adoc:33]:
```suggestion
- *History*: prevent the usage of old passwords.
```

File: pages/am/3.x/user-guide/user-management/password/user-guide-password-policy.adoc [hunk 15]
```diff
@@ -23,12 +23,14 @@ NOTE: You can also configure password policy settings at security domain level,
...
 You can set the following password characteristics:
 
-- *Minimum length*: length of the passwords should be enforced by the application.
+- *Minimum & Maximum length*: length of the passwords should be enforced by the application.
 - *Numbers*: must include at least one number.
+- *Expiry Duration*: the expiration duration (in days) of a password.
 - *Special characters*: must include at least one special character.
 - *Mixed case*: must include lower and upper case letters.
 - *Exclude common passwords*: will exclude common passwords from a dictionary.
 - *Exclude user profile information from passwords*: will user profile information within passwords (case insensitive).
```
Comment (ivan-gravitee) [inline pages/am/3.x/user-guide/user-management/password/user-guide-password-policy.adoc:32]:
```suggestion
- *Exclude user profile information from passwords*: will exclude user profile information from use in passwords (case insensitive).
```

**Review by ivan-gravitee — CHANGES_REQUESTED** [review]
Hey @leleueri , I've reviewed the files and made a few suggestions, let me know if you have questions about any of these.

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+  mfa_rate:
+    enabled: true
+    limit: 5
+    timePeriod: 15
+    timeUnit: Minutes
+----
+
+Gravitee AM monitors the MFA challenge request based on the enabled rate limit configuration and shows the user rate limit exceeded message if the user exceeds the limit.
+Then the user has to wait a certain amount of time before making a successful request.
+For an example; if the limit is set to 2 for 1 minute time period and the user already has sent 2 requests, then the user must wait another 30 seconds before being able to send the 3rd request.
```
[the remark this answers is not in the document]
Comment (leleueri) [reply pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:50]: @ashraf706 what is your view ?

File: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc (hunk 5 above)
Comment (leleueri) [reply pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc:61]: yes, exactly the same screen at application level than at domain level.

**Review by ivan-gravitee — APPROVED** [review]
LGTM

**Review by ivan-gravitee — APPROVED** [review]
LGTM - only one pending comment [, do we need this addressed first?](https://github.com/gravitee-io/gravitee-docs/pull/948/files#r1061312012) but we can address this at a later point as well.

**Comment by leleueri** [comment]
> LGTM - only one pending comment [, do we need this addressed first?](https://github.com/gravitee-io/gravitee-docs/pull/948/files#r1061312012) but we can address this at a later point as well.

We have accepted the suggestion.

**Pull request merged**